spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/app/AbstractActuatorTemplate.java

```diff
@@ -66,55 +82,80 @@ protected AbstractActuatorTemplate(RestTemplate restTemplate, AppDeployer appDep
 
 	@Override
 	public <T, R> R postToActuator(String deploymentId, String guid, String endpoint, T body,
-			Class<R> responseType) {
-		String actuatorUrl = getActuatorUrl(deploymentId, guid);
+			Class<R> responseType, Optional<HttpHeaders> optionalRequestHeaders) {
+		AppInstanceStatus appInstanceStatus = getDeployedInstance(deploymentId, guid)
+				.orElseThrow(() -> new IllegalStateException(
+						String.format("App with deploymentId %s and guid %s not deployed", deploymentId, guid)));
+
+		String actuatorUrl = getActuatorUrl(appInstanceStatus);
+
+		HttpHeaders requestHeaders = requestHeaders(httpHeadersForInstance(appInstanceStatus), optionalRequestHeaders);
 
 		ResponseEntity<R> responseEntity = httpPost(UriComponentsBuilder
-				.fromHttpUrl(actuatorUrl).path(normalizePath(endpoint)).toUriString(), body, responseType);
+				.fromHttpUrl(actuatorUrl).path(normalizePath(endpoint)).toUriString(), body, responseType,
+				requestHeaders);
 		if (responseEntity.getStatusCode().isError()) {
 			logger.error(responseEntity.getStatusCode().toString());
 		}
 		return responseEntity.getBody();
 	}
 
-	protected final HttpHeaders requestHeaders() {
-		HttpHeaders headers = new HttpHeaders();
-		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
-		headers.setContentType(MediaType.APPLICATION_JSON);
-		this.httpHeadersConsumer.accept(headers);
-		return headers;
+	protected final String getActuatorUrl(AppInstanceStatus appInstanceStatus) {
+		try {
+			return actuatorUrlForInstance(appInstanceStatus);
+		} catch (Exception e) {
+			throw new IllegalArgumentException(String.format(
+					"Unable to determine actuator url for app with guid %s",
+					appInstanceStatus.getAttributes().get("guid")));
+		}
 	}
 
-	protected final void setHttpHeadersConsumer(Consumer<HttpHeaders> httpHeadersConsumer) {
-		Assert.notNull(httpHeadersConsumer, "'httpHeadersConsumer cannot be null");
-		this.httpHeadersConsumer = httpHeadersConsumer;
+	protected abstract String actuatorUrlForInstance(AppInstanceStatus appInstanceStatus);
+
+	/**
+	 * Hook to allow subclasses to add special headers derived {@link AppInstanceStatus} metadata if necessary.
+	 *
+	 * @param appInstanceStatus the AppInstanceStatus for the target instance.
+	 * @return HttpHeaders
+	 */
+	protected Optional<HttpHeaders> httpHeadersForInstance(AppInstanceStatus appInstanceStatus) {
+		return Optional.empty();
 	}
 
-	private final String normalizePath(String path) {
-		return path.startsWith("/") ? path : "/" + path;
+
+	private final HttpHeaders requestHeaders(Optional<HttpHeaders> optionalAppInstanceHeaders,
+			Optional<HttpHeaders> optionalRequestHeaders) {
+		HttpHeaders requestHeaders = optionalAppInstanceHeaders
+				.orElse(new HttpHeaders());
+		optionalRequestHeaders.ifPresent(requestHeaders::addAll);
```

Comment:
Do we need to update any tests to get coverage of these header overrides etc? 

Same for the "request headers already contains auth header" scenario.